File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/sparse/Spdiags.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.sparse;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Builds an {@code n × n} diagonal CSR sparse matrix from a 1D diagonal vector.
 *
 * <p>C++ op name: {@code spdiags}
 *
 * <p><b>Input (1 array):</b>
 * <ol>
 *   <li>{@code diag} – 1D [n], floating dtype — the diagonal values</li>
 * </ol>
 *
 * <p><b>Integer argument (IArg):</b> {@code n} — size of the resulting square matrix.
 *
 * <p><b>Outputs (3 arrays, CSR of the n×n diagonal matrix):</b>
 * <ol>
 *   <li>{@code values}  – 1D [n], same dtype as {@code diag} — the non-zero values
 *       (identical to {@code diag})</li>
 *   <li>{@code colIdx}  – 1D [n], INT32 — column indices {@code [0, 1, ..., n-1]}</li>
 *   <li>{@code rowPtr}  – 1D [n+1], INT32 — row pointers {@code [0, 1, ..., n]}</li>
 * </ol>
 *
 * <p><b>Autodiff:</b> fully differentiable — gradient flows directly to {@code diag}
 * from the upstream gradient of the {@code values} output, because
 * {@code values[k] == diag[k]} identically.
 */
public class Spdiags extends DynamicCustomOp {

    /** No-arg constructor required for op-registry reflection. */
    public Spdiags() {}

    /**
     * Eager (INDArray) constructor.
     *
     * @param diag 1D [n] diagonal values (floating dtype)
     * @param n    size of the square output matrix
     */
    public Spdiags(INDArray diag, long n) {
        super(new INDArray[]{diag}, null);
        addIArgument(n);
    }

    /**
     * SameDiff (symbolic) constructor.
     *
     * @param sd   the SameDiff graph
     * @param diag SDVariable [n] diagonal values
     * @param n    size of the square output matrix
     */
    public Spdiags(SameDiff sd, SDVariable diag, long n) {
        super(sd, new SDVariable[]{diag});
        addIArgument(n);
    }

    @Override
    public String opName() {
        return "spdiags";
    }

    /**
     * Output data types:
     * <ol>
     *   <li>values  – same dtype as {@code diag} (input index 0)</li>
     *   <li>colIdx  – INT32</li>
     *   <li>rowPtr  – INT32</li>
     * </ol>
     */
    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        return Arrays.asList(dataTypes.get(0), DataType.INT32, DataType.INT32);
    }

    /**
     * Backward pass for spdiags.
     *
     * <p>Because {@code values[k] == diag[k]} exactly, the gradient of the loss with
     * respect to {@code diag} is simply the upstream gradient of the {@code values} output:
     * <pre>
     *   d(loss)/d(diag) = grads.get(0)   (upstream grad of values)
     * </pre>
     * {@code colIdx} and {@code rowPtr} are integer/structural and carry no gradient.
     *
     * @param grads upstream gradients; {@code grads.get(0)} is the gradient w.r.t. values output
     * @return list with a single element: the gradient w.r.t. {@code diag}
     */
    @Override
    public List<SDVariable> doDiff(List<SDVariable> grads) {
        // grads has one entry per output: [d_values, d_colIdx, d_rowPtr]
        // d_colIdx and d_rowPtr are always zero (structural INT32 outputs).
        // d(diag) = d_values because values IS the diag.
        return Collections.singletonList(grads.get(0));
    }
}

File: libnd4j/include/ops/declarable/generic/sparse/spdiags.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// spdiags — build a diagonal CSR from a vector.
//
// Given a float vector diag[n], produces the CSR representation of the n×n
// diagonal matrix with diag on the main diagonal.  This is a trivial
// (data-independent) shape operation:
//   nnz = n,  rowPtr[i] = i,  colIdx[i] = i,  values[i] = diag[i].
//
// Input:
//   [0] diag  [n], floating dtype
// IArgs:
//   [0] n  — size of the diagonal (and of the square matrix)
// Outputs:
//   [0] values  [n]   — same dtype as diag  (= copy of diag)
//   [1] colIdx  [n]   — INT32, colIdx[i] = i
//   [2] rowPtr  [n+1] — INT32, rowPtr[i] = i
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_spdiags)

#include <ops/declarable/headers/sparse.h>
#include <ops/declarable/helpers/sparse_spdiags.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(spdiags, 1, 3, false, 0, 1) {
  auto diag = INPUT_VARIABLE(0);

  const LongType n = INT_ARG(0);

  REQUIRE_TRUE(diag->rankOf() == 1, 0,
               "spdiags: diag must be 1D, got rank %d", diag->rankOf());
  REQUIRE_TRUE(diag->lengthOf() >= n, 0,
               "spdiags: diag length %lld must be >= n=%lld",
               (long long)diag->lengthOf(), (long long)n);
  REQUIRE_TRUE(n >= 0, 0, "spdiags: n must be non-negative, got %lld", (long long)n);

  auto values = OUTPUT_VARIABLE(0);
  auto colIdx = OUTPUT_VARIABLE(1);
  auto rowPtr = OUTPUT_VARIABLE(2);

  sd::ops::helpers::spdiags(block.launchContext(), *diag, *values, *colIdx, *rowPtr, n);

  return sd::Status::OK;
}

DECLARE_SHAPE_FN(spdiags) {
  auto diag = INPUT_VARIABLE(0);

  const LongType n       = INT_ARG(0);
  const auto     dtype   = diag->dataType();
  const auto     idxType = sd::DataType::INT32;

  auto valShape = ConstantShapeHelper::getInstance().createShapeInfo(dtype,   'c', {n});
  auto ciShape  = ConstantShapeHelper::getInstance().createShapeInfo(idxType, 'c', {n});
  auto rpShape  = ConstantShapeHelper::getInstance().createShapeInfo(idxType, 'c', {n + 1});

  return SHAPELIST(valShape, ciShape, rpShape);
}

DECLARE_TYPES(spdiags) {
  getOpDescriptor()
      ->setAllowedInputTypes(0,  {ALL_FLOATS})  // diag
      ->setAllowedOutputTypes(0, {ALL_FLOATS})  // values
      ->setAllowedOutputTypes(1, {ALL_INTS})    // colIdx
      ->setAllowedOutputTypes(2, {ALL_INTS});   // rowPtr
  getOpDescriptor()->addTraits(OP_TRAIT_DATA_MOVEMENT | OP_TRAIT_FULLY_WRITING);
}

}  // namespace ops
}  // namespace sd

#endif  // NOT_EXCLUDED(OP_spdiags)
